### DISCUSSION TITLE
How to create an EC Public key using affine coordinates and non-deprecated calls

### DISCUSSION THREAD

**Discussion by ldts** [openssl/openssl#21872] [Q&A] [answered]
Some criptoIPs can generate the EC PUB_X/PUB_Y affine coordinates on EC curves (their private keys never accessible outside the silicon)

These cryptographic IPs can be under control of a Trusted Execution Environment (see)
https://github.com/OP-TEE/optee_client/pull/352 and the TEE at https://github.com/OP-TEE/optee_os/blob/master/core/pta/imx/manufacturing_protection.c 

The following code, allows the user to validate the values read from the cryptographic IP and convert the public key to DER or PEM format

https://github.com/foundriesio/lmp-device-register/commit/e763be5eae42d7c3222e1c7e341f16bc8c8b992c#diff-2d0eeb1947d28eea137eceb59a6a329480209c4848d432396bcb6c1fe49fd327R189

The issue with the code is that it uses many deprecated calls.

I have been trying to rewrite the function using EVP_KEY but I believe there is a regression:

```
static int foo(string &raw)
{
	EVP_PKEY_CTX *genctx = NULL;
	OSSL_LIB_CTX *libctx = NULL;
	const char *propq = NULL;
	EVP_PKEY *key = NULL;

	genctx = EVP_PKEY_CTX_new_from_name(libctx, "EC", propq);
	
	if (genctx == NULL)
		leave;

	if (EVP_PKEY_keygen_init(genctx) <= 0)
		leave;

	BIGNUM *x_bn = nullptr;
	BIGNUM *y_bn = nullptr;
	
	x_bn = BN_bin2bn((unsigned char *)raw.c_str(),
			raw.size() / 2, nullptr);

	y_bn = BN_bin2bn((unsigned char *)raw.c_str() + raw.size() / 2,
			raw.size() / 2, nullptr);

	OSSL_PARAM_BLD *bld = OSSL_PARAM_BLD_new();

	OSSL_PARAM_BLD_push_utf8_string(bld, OSSL_PKEY_PARAM_GROUP_NAME,
					(char *) SN_X9_62_prime256v1, 0);

	OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_EC_PUB_X, x_bn);

	OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_EC_PUB_Y, y_bn);

	OSSL_PARAM *params = OSSL_PARAM_BLD_to_param(bld);

	if (!EVP_PKEY_CTX_set_params(genctx, params))
		leave;

	if (EVP_PKEY_generate(genctx, &key) <= 0)
		leave;
}
```

Keys generated using this method above will _overwrite_ the PUB_X/PUB_Y values without generating any errors.
I believe this to be a regression unless there is an interface that I am not aware for this use case.

From the documentation it seems that OSSL_PKEY_PARAM_EC_PUB_X and OSSL_PKEY_PARAM_EC_PUB_Y can only be used to read those values back - not to set them.

If anyone could help either by accepting the bug report or proposing an alternative that would be much appreciated .

**Comment by t8m** [comment, marked as answer]
Please see the second example at the https://www.openssl.org/docs/man3.0/man3/EVP_PKEY_fromdata.html manual page. To create just a public key instead of a keypair you use EVP_PKEY_PUBLIC_KEY as the selection and of course you don't set the priv param.

**Comment by ldts** [comment]
argh, sorry, not sure how I didnt find this. my bad.
thanks a lot for your help.
